File: solutions/typescript/grade-school/1/grade-school.ts
export class GradeSchool {
  private _students = new Map<string, number>();
  private _roster: Record<number, string[]> = {};

  // O(MN), M -> properties in _roster, N -> largest array in _roster 
  roster(): Record<number, string[]> {
    const record: Record<number, string[]> = {};
    for (const key of Object.keys(this._roster)) {
      const grade = parseInt(key)
      record[grade] = this._roster[grade]?.slice(0);
    }
    return record;
  }

  // O(NLGN)
  add(name: string, grade: number): void {
    const oldGrade: number | undefined = this._students.get(name);
    if (oldGrade) {
      this._roster[oldGrade] = this._roster[oldGrade].filter(s => s !== name);
      if (this._roster[oldGrade].length === 0)
        delete this._roster[oldGrade];
    }
    this._students.set(name, grade);

    if (!(grade in this._roster)) {
      this._roster[grade] = [name];
    } else {
      this._roster[grade].push(name)
      this._roster[grade].sort();
    }
  }

  // O(N), N -> this._roster[grade].length
  grade(grade: number): string[] {
    return this._roster[grade] ? this._roster[grade].slice(0) : [];
  }
}

File: solutions/cpp/grade-school/1/grade_school.cpp
#include "grade_school.h"
#include <algorithm>
#include <vector>

namespace grade_school {
school::school() : _roster() {}

const std::map<int, std::vector<std::string>> school::roster() const {
  return _roster;
}

void school::add(std::string name, int grade) {
  if (_roster.count(grade) > 0) {
    _roster[grade].push_back(name);
    std::sort(_roster.at(grade).begin(), _roster.at(grade).end());
  } else {
    _roster.insert({grade, std::vector<std::string>{name}});
  }
}
std::vector<std::string> school::grade(int grade) const {
  if (_roster.count(grade) > 0) {
    return _roster.at(grade);
  }
  return std::vector<std::string>{};
}

} // namespace grade_school
